qiskit/transpiler/passes/crosstalk_adaptive_schedule.py

```diff
@@ -0,0 +1,667 @@
+                    for tmpg in off_set:
+                        clauses.append(Not(self.overlap_indicator[gate][tmpg]))
+                    err = 0
+                    if not on_set:
+                        err = self.bp_cx_err[self.cx_tuple(gate)]
+                    elif len(on_set) == 1:
+                        on_gate = on_set[0]
+                        err = self.crosstalk_prop[self.gate_tuple(gate)][self.gate_tuple(on_gate)]
+                    else:
+                        err_list = []
+                        for on_gate in on_set:
+                            tmp_prop = self.crosstalk_prop[self.gate_tuple(gate)]
+                            err_list.append(tmp_prop[self.gate_tuple(on_gate)])
+                        err = max(err_list)
+                    if err == 1.0:
+                        err = 0.999999
+                    val = round(math.log(1.0 - err), NUM_PREC)
+                    self.opt.add(Implies(And(*clauses), self.gate_fidelity[gate] == val))
+
+    def coherence_constraints(self):
+        """
+        Set decoherence errors based on qubit lifetimes
+        """
+        self.last_gate_on_qubit = {}
+        for gate in self.dag.topological_op_nodes():
+            if isinstance(gate.op, Measure):
+                continue
+            if isinstance(gate.op, Barrier):
+                continue
+            if len(gate.qargs) == 1:
+                q_0 = gate.qargs[0].index
+                self.last_gate_on_qubit[q_0] = gate
+            else:
+                q_0 = gate.qargs[0].index
+                q_1 = gate.qargs[1].index
+                self.last_gate_on_qubit[q_0] = gate
+                self.last_gate_on_qubit[q_1] = gate
+
+        self.first_gate_on_qubit = {}
+        for gate in self.dag.topological_op_nodes():
+            if len(gate.qargs) == 1:
+                q_0 = gate.qargs[0].index
+                if q_0 not in self.first_gate_on_qubit:
+                    self.first_gate_on_qubit[q_0] = gate
+            else:
+                q_0 = gate.qargs[0].index
+                q_1 = gate.qargs[1].index
+                if q_0 not in self.first_gate_on_qubit:
+                    self.first_gate_on_qubit[q_0] = gate
+                if q_1 not in self.first_gate_on_qubit:
+                    self.first_gate_on_qubit[q_1] = gate
+
+        for q in self.last_gate_on_qubit:
+            g_last = self.last_gate_on_qubit[q]
+            g_first = self.first_gate_on_qubit[q]
+            finish_time = self.gate_start_time[g_last] + self.gate_duration[g_last]
+            start_time = self.gate_start_time[g_first]
+            if q in self.measured_qubits:
+                self.opt.add(self.measure_start >= finish_time)
+                self.opt.add(self.qubit_lifetime[q] == self.measure_start - start_time)
+            else:
+                # All qubits get measured simultaneously whether or not they need a measurement
```

Comment:
can you also mention this assumption in the pass docstring for clarity? it's likely that this can be relaxed in future generation devices.

Reply:
Added